cnf-certification-test/networking/suite.go

```diff
@@ -67,8 +80,106 @@ var _ = ginkgo.Describe(common.NetworkingTestKey, func() {
+	ginkgo.It(testID, ginkgo.Label(testID), func() {
+		testListenAndDeclared(&env)
+	})
 })
 
+func parseListening(res string, listeningPorts map[key]string) {
+	var k key
+	lines := strings.Split(res, "\n")
+	for _, line := range lines {
+		fields := strings.Fields(line)
+		if !strings.Contains(line, "LISTEN") {
+			continue
+		}
+		if indexprotocolname > len(fields) || indexport > len(fields) {
+			return
+		}
+		s := strings.Split(fields[indexport], ":")
+		p, _ := strconv.Atoi(s[1])
+		k.port = p
+		k.protocol = strings.ToUpper(fields[indexprotocolname])
+		k.protocol = strings.ReplaceAll(k.protocol, "\"", "")
+		listeningPorts[k] = ""
+	}
+}
+
+func checkIfListenIsDeclared(listeningPorts, declaredPorts map[key]string) map[key]string {
+	res := make(map[key]string)
+	if len(listeningPorts) == 0 {
+		return res
+	}
+	for k := range listeningPorts {
+		_, ok := declaredPorts[k]
+		if !ok {
+			tnf.ClaimFilePrintf(fmt.Sprintf("The port %d on protocol %s in pod %s is not declared.", k.port, k.protocol, listeningPorts[k]))
+			res[k] = listeningPorts[k]
+		}
+	}
+	return res
+}
+
+//nolint:funlen
+func testListenAndDeclared(env *provider.TestEnvironment) {
+	var k key
+	declaredPorts := make(map[key]string)
+	listeningPorts := make(map[key]string)
+	var failedPods string
+	var skippedPods string
+
+	for _, pod := range env.Pods {
+		for i := 0; i < len(pod.Spec.Containers); i++ {
+			container := &pod.Spec.Containers[i]
+			ports := container.Ports
+			fmt.Println(ports)
+			if ports == nil {
+				tnf.ClaimFilePrintf("Failed to get declared port for container %d due to %v, skipping pod %s", container, pod.Namespace+"."+pod.Name)
+				skippedPods += pod.Namespace + " " + pod.Name + " " + container.Name + "\n"
+			}
+			for j := 0; j < len(ports); j++ {
+				k.port = int(ports[j].ContainerPort)
+				k.protocol = string(ports[j].Protocol)
+				declaredPorts[k] = ports[j].Name
+			}
+
+			oc := clientsholder.GetClientsHolder()
+			output, outerr, err := oc.ExecCommandContainer(clientsholder.Context{Namespace: pod.Namespace,
+				Podname: pod.Name, Containername: pod.Spec.Containers[i].Name}, `ss -tulwnH`)
+			if err != nil {
+				logrus.Errorln("can't execute command on container ", err)
+				continue
+			}
+			if outerr != "" {
+				logrus.Errorln("error when running listening command ", outerr)
+				continue
+			}
+			parseListening(output, listeningPorts)
+			if len(listeningPorts) == 0 {
+				tnf.ClaimFilePrintf("Failed to get listening port for pod name %s in pod namespace %s, skipping this pod", pod.Name, pod.Namespace, err)
+				continue
+			}
+			// compare between declaredPort,listeningPort
+			undeclaredPorts := checkIfListenIsDeclared(listeningPorts, declaredPorts)
+			for k := range undeclaredPorts {
+				tnf.ClaimFilePrintf("The port %d on protocol %s in pod name %s and pod namespace is %s not declared.", k.port, k.protocol, pod.Name, pod.Namespace)
```

Comment:
you can use the PodToString() function to display the pod

Reply:
I changed.